File: Week-2/2-vigenere/vigenere.c
#include <cs50.h>
#include <stdio.h>
#include <ctype.h>
#include <string.h>

int shift(char c);

int main(int argc, string argv[])
{
    if (argc != 2)
    {
        printf("Usage: ./vigenere keyword\n");
        return 1;
    }
    // Validate key
    string s = argv[1];
    int i = 0;
    char c = s[i];  
    while (c != '\0')
    {
        if (!isalpha(c))
        {
            printf("Usage: ./vigenere keyword\n");
            return 1;
        }
        c = s[++i];
    }
    int keylen = strlen(s);
    int k = 0;
    // Ask user for a plain text
    string plain = get_string("plaintext: ");
    // Encrypt plain text using user key
    c = plain[0];
    i = 0;
    printf("ciphertext: ");
    while (c != '\0')
    {
        // Loop key
        if (k >= keylen)
        {
            k = 0;
        }
        int key = shift(s[k]);
        // Cypher
        char ck = c;
        if (c >= 'a' && c <= 'z')
        {
            // Check lower case
            ck = (c + key) > 'z' ? 'a' + (c + key - 1) - 'z' : (c + key);
            k++;
        } 
        else if (c >= 'A' && c <= 'Z')
        {
            // Check upper case
            ck = (c + key) > 'Z' ? 'A' + (c + key - 1) - 'Z' : (c + key);
            k++;
        } 
        printf("%c", ck);
        // Pick next char
        c = plain[++i];
    }
    printf("\n");
}

// Convert character into the correct shift value
int shift(char c)
{
    if (c >= 'a' && c <= 'z')
    {
        // Check lower case
        return c - 'a';
    } 
    else if (c >= 'A' && c <= 'Z')
    {
        // Check upper case
        return c - 'A';
    }
    return c;
}

File: Week-6/7 - vigenere/vigenere.py
from sys import argv, exit
from cs50 import get_string

# Key length
kl = 0
# Key pointer
kp = 0


def encrypt(c, k):
    global kl
    global kp
    code = ord(c)
    # Check non alphabetic symbol
    if not ((code >= 65 and code <= 90) or (code >= 97 and code <= 122)):
        return str(chr(code))
    shift = code
    # Get key code
    key_code = ord(k[kp])
    # Move pointer
    kp += 1
    if kp >= kl:
        kp = 0
    # Calculate shift value
    if key_code >= 65 and key_code <= 90:
        # Uppercase key
        shift += key_code - 65
    elif key_code >= 97 and key_code <= 122:
        # Lowercase key
        shift += key_code - 97
    # Encrypt
    if code >= 65 and code <= 90:
        # Uppercase
        return str(chr(shift)) if shift <= 90 else str(chr(65 + (shift - 65) % 26))
    elif code >= 97 and code <= 122:
        # Lowercase
        return str(chr(shift)) if shift <= 122 else str(chr(97 + (shift - 97) % 26))
    else:
        return str(chr(code))


# Validate argv
if len(argv) != 2:
    print("Usage: python vigenere.py k")
    exit(1)
# Read command line arg
k = argv[1]
# Validate key
if any(char.isdigit() for char in k):
    print("invalid keyword")
    exit(1)
# Ask user for text
text = get_string("plaintext:  ")
kl = len(k)
# Encrypt text
encrypted = ""
for c in text:
    encrypted += encrypt(c, k)
# Print encrypted text
print(f"ciphertext: {encrypted}")